x-pack/elastic-agent/pkg/agent/application/filters/stream_checker.go

```diff
@@ -0,0 +1,184 @@
+							}
+						}
+					}
+				}
+			}
+		}
+
+		streamsNode, ok := inputNode.Find("streams")
+		if !ok {
+			continue
+		}
+
+		streamsList, ok := streamsNode.Value().(*transpiler.List)
+		if !ok {
+			continue
+		}
+
+		streamNodes, ok := streamsList.Value().([]transpiler.Node)
+		if !ok {
+			return errors.New("streams is not a list", errors.TypeConfig)
+		}
+
+		for _, streamNode := range streamNodes {
+			streamMap, ok := streamNode.(*transpiler.Dict)
+			if !ok {
+				continue
+			}
+
+			// fix this only if in compact form
+			if dsNameNode, found := streamMap.Find("dataset.name"); found {
+				dsKey, ok := dsNameNode.(*transpiler.Key)
+				if ok {
+					if newDataset := dsKey.Value().(transpiler.Node).String(); !isDatasetValid(newDataset) {
+						return ErrInvalidDataset
+					}
+				}
+			} else {
+				datasetNode, found := streamMap.Find("dataset")
+				if found {
+					datasetMap, ok := datasetNode.Value().(*transpiler.Dict)
+					if !ok {
+						continue
+					}
+
+					dsNameNode, found := datasetMap.Find("name")
+					if found {
+						dsKey, ok := dsNameNode.(*transpiler.Key)
+						if ok {
+							if newDataset := dsKey.Value().(transpiler.Node).String(); !isDatasetValid(newDataset) {
+								return ErrInvalidDataset
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// The only two requirement are that it has only characters allowed in an Elasticsearch index name
+// and does NOT contain a `-`.
+// Index names must meet the following criteria:
+//     Lowercase only
+//     Cannot include \, /, *, ?, ", <, >, |, ` ` (space character), ,, #
+//     Cannot start with -, _, +
+//     Cannot be . or ..
+
+func isNamespaceValid(namespace string) bool {
+	// Cannot be . or ..
+	if namespace == "." || namespace == ".." {
+		return false
+	}
+
+	if len(namespace) <= 0 || len(namespace) > 255 {
+		return false
+	}
+
+	// Lowercase only
+	if strings.ToLower(namespace) != namespace {
+		return false
+	}
+
+	// Cannot include \, /, *, ?, ", <, >, |, ` ` (space character), ,, #
+	if strings.ContainsAny(namespace, "\\/*?\"<>| ,#-") {
+		return false
+	}
+
+	// Cannot start with -, _, +
+	if strings.HasPrefix(namespace, "-") || strings.HasPrefix(namespace, "_") || strings.HasPrefix(namespace, "+") {
+		return false
+	}
+
+	return true
+}
+
+// The same requirements as for the namespace apply.
+func isDatasetValid(dataset string) bool {
+	// Cannot be . or ..
+	if dataset == "." || dataset == ".." {
+		return false
+	}
+
+	if len(dataset) <= 0 || len(dataset) > 255 {
+		return false
+	}
+
+	// Lowercase only
+	if strings.ToLower(dataset) != dataset {
+		return false
+	}
+
+	// Cannot include \, /, *, ?, ", <, >, |, ` ` (space character), ,, #
+	if strings.ContainsAny(dataset, "\\/*?\"<>| ,#-") {
+		return false
+	}
+
+	// Cannot start with -, _, +
+	if strings.HasPrefix(dataset, "-") || strings.HasPrefix(dataset, "_") || strings.HasPrefix(dataset, "+") {
+		return false
+	}
+
+	return true
```

Comment:
Why duplicate the same function and just give it the same name? Could remove the second one and rename it to `isValid`. Or could do:

```
func isDatasetValid(dataset string) bool {
    return isNamespaceValid(dataset)
}
```

That way in the future if the validation of dataset does differ from namespace, then only that function body needs to be changed.